### ISSUE TITLE
GroupBy sum count not refreshing after row delete

### ISSUE THREAD

**Issue by tsbalaji69** [free-jqgrid/jqGrid#452]
Hi,
When I delete row in a group how to recalculate the sum and count and also how to get the sum and count value displayed in the header?
Regards,
balaji
Screen shot
![image](https://user-images.githubusercontent.com/43889770/46549529-fe0ffe00-c8ef-11e8-9145-d4aa4e8b69b2.png)

**Comment by OlegKi** [comment]
Could you post JavaScript code, which would show the options of jqGrid, which you use? Do you use client side grouping or server side grouping?

In the most cases the simplest way to update grouping headers would be reloading the grid with respect of `.trigger("reloadGrid")` with some additional parameters, which could depend on your configuration.

**Comment by tsbalaji69** [comment]
Hi OlegKi,
Thank you for the response. I need to load data from server only once. User will delete rows and the sum / count should re calculate accordingly.
My Java script code related to Jqgrid are
`<script>
	$.jgrid = $.jgrid || {};
	$.jgrid.no_legacy_api = false;
</script>`
To populate jqgrid
` //=========================================
    $("#dataGrid").jqGrid({  // Load jqgrid with data
        url: 'displaydata.aspx/GetDataFromDB',
        datatype: 'json',
        mtype: 'POST',
        serializeGridData: function (postData) {
            return JSON.stringify(postData);
        },
        ajaxGridOptions: { contentType: "application/json" },
        loadonce: true,
        guiStyle: "bootstrap",
        iconSet: "fontAwesome",
        colNames: ['Ship From', 'Ship Date', 'Order Number', 'PO#', 'Shipment ID#', 'Intl DRP CODE', '#SKU', 'Estimated Arrival Date', 'Quantity'],
        colModel: [

                    { name: 'TRLOC', index: 'TRLOC', width: 160 },
                    { name: 'TRSHPDT', index: 'TRSHPDT', width: 100 },
                    { name: 'TRORD', index: 'TRORD', width: 100 },
                    { name: 'TRCSTPO', index: 'TRCSTPO', width: 100 },
                    { name: 'TRSHPID', index: 'TRSHPID', width: 100 },
                    { name: 'TRDRPC', index: 'TRDRPC', width: 100 },
                    {
                        name: 'SKU#', index: 'SKU#', width: 60, align: 'center', sortable: false, summaryType: 'count', summaryTpl: '<b>{0}</b>'
                    },
                    { name: 'TRARDT', index: 'TRARDT', width: 100 },
                    {
                        name: 'TROSQTY', index: 'TROSQTY', width: 60, align: 'center', sortable: false, formatter: "number", summaryType: 'sum'
                    },

        ],
        pager: '#pagingGrid',
        //rowNum: 18,  
        //rowList: [30, 60, 90],  
        viewrecords: true,
        gridview: true,
        autowidth: true,
        jsonReader: {
            page: function (obj) { return 1; },
            total: function (obj) { return 1; },
            records: function (obj) { return obj.d.length; },
            root: function (obj) { return obj.d; },
            repeatitems: false,
            id: "0"
        },
        grouping: true,
        multiselect: true,
        multiselectPosition: "right",
        multiselectWidth: 80,
        multiselectText: "Received",
        groupCollapse: true,
        groupingView: {
            groupField: ['TRLOC', 'TRORD'],
            
            groupText: ['Ship From : {0}', 'OrdNum: {0} <input type="checkbox" class="groupHeader"/>'],
            // groupColumnShow : [false],
            groupSummary: [true, true],
            groupSummaryPos: ['header', 'header'],
            groupOrder: ['asc', 'asc']

        },
        onSelectAll: function (rowIds, allChecked) {
            $("input.groupHeader").attr('checked', allChecked);
        },
        caption: 'Order Processing Data'
    });`
On a custom button click i try to get the sum and count values ,delete selected rows and recalculate sum and count for the existing rows
`    $('#btnProcess').click(function (e) {
        e.preventDefault();
        var selectedrows = $("#dataGrid").jqGrid('getGridParam', 'selarrrow');
        var datatosend = [];
        if (selectedrows.length) {
            for (var i = 0; i < selectedrows.length; i++) {
                var selectedrowdetails = $("#dataGrid").jqGrid('getRowData', selectedrows[i]);
                datatosend.push(selectedrowdetails);
            }
            //remove selected rows
            for (var i = selectedrows.length - 1; i >= 0; i--) {
                $("#dataGrid").jqGrid('delRowData', selectedrows[i], { reloadAfterSubmit: true });
            }
        }
        //Get the values of sum and count and Re calcualte sum and count of the group header
        // jQuery("#dataGrid").jqGrid('groupingRemove');
        //jQuery("#dataGrid").jqGrid('groupingGroupBy', 'data');
        // jQuery("#dataGrid").jqGrid('getGridParam', 'data');
        //  $("#dataGrid").jqGrid('getRowData');
       
    });`
Kindly can you guide me in getting the sum and count values present at the headers and re calculate it after row delete.
Thank you once again for the concern,
regards,
Balaji

**Comment by tsbalaji69** [comment]
Hi OleK,
I formatted the code and updated the comments. Please do help.
Thank you,
regards,
Balaji

**Comment by OlegKi** [comment]
As I mention before I think that the most easy way to recalculate grouping summary would be just to reload the grid *locally* (I mean that one don't need to reload the data from the server). To do this one need to add `$("#dataGrid").trigger("reloadGrid", { page: 1 });` (or just `$("#dataGrid").trigger("reloadGrid");`) after deleting the rows:
```JavaScript
$('#btnProcess').click(function (e) {
    var $grid = $("#dataGrid"), i, datatosend = [], selectedrowdetails,
        selectedrows = $grid.jqGrid('getGridParam', 'selarrrow');

    e.preventDefault();
    if (selectedrows.length) {
        for (i = 0; i < selectedrows.length; i++) {
            selectedrowdetails = $grid.jqGrid('getRowData', selectedrows[i]);
            datatosend.push(selectedrowdetails);
        }
        //remove selected rows
        for (i = selectedrows.length - 1; i >= 0; i--) {
            $grid.jqGrid('delRowData', selectedrows[i]);
        }
        $grid.trigger("reloadGrid", { page: 1 });
    }
});
```
See the demo https://jsfiddle.net/OlegKi/72vsjurq/
If I correctly understand your problem then it should solve your problem.

**Comment by tsbalaji69** [comment]
Hi OlegKi,
Thank you so much as the reload works as intended.Kindly can you guide me in getting the sum and count values present at the headers? I am able to get the row values but unable to get the sum and count present in the header.
Thank you,
regards,
balaji

**Comment by OlegKi** [comment]
Sorry, but I don't understand your question. One can see on the picture of your first post and on the demo, which I posted that both the sum and the count will be show in the headers. Thus there already exist in the header because of usage `summaryType: 'count'` or `summaryType: 'sum'` in `colModel`. It's unclear for me which problem you still have?

**Comment by tsbalaji69** [comment]
Hi OlegKi,
Yes I have them in the header. I need to retrieve them on the delete button click. I can get the row values using 
```
for (i = 0; i < selectedrows.length; i++) {
                selectedrowdetails = $grid.jqGrid('getRowData',
                        selectedrows[i]);
                datatosend.push(selectedrowdetails);
            }
```
When i try to use the below code it returns null
```
 var groupHeadersOptions = $("#dataGrid").jqGrid("getGridParam", "groupHeader");
```
In the below screen shot
![image](https://user-images.githubusercontent.com/43889770/46622294-a8359300-cb47-11e8-8113-ef789524e3ee.png)

I need to get the value 7 and 11.00 which is the summaryType: count and summaryType: sum

Heart felt thanks for your patience in understanding my explanation and helping me out.
regards,
Balaji

**Comment by OlegKi** [comment]
You can use 
```JavaScript
var groupingView = $("#dataGrid").jqGrid("getGridParam", "groupingView"),
    groups = groupingView.groups;
```
to get grouping information. The values `groups[i].keys`, `groups[i].summary[j].nm`, `groups[i].summary[j].v` will get you full information about the grouping values and the summary values. The number of items in `groups` corresponds the number of grouping headers. You can get only the latest groups by filtering the `groups` array:
```JavaScript
groups = groups.filter(function (group) {
    return group.keys.length === groupingView.groupField.length;
});
```
I hope it's the information, which you missed.

**Comment by tsbalaji69** [comment]
Hi OlegKi,
I am really moved by your helping tendency. My heart felt thank you for helping me out.
Regards,
Balaji

**Comment by tsbalaji69** [comment]
Hi OlegKi,
Thank you for the help. I will close this issue and will open a new issue for my requirement which is extension of my earlier grid question.
Regards,
balaji

**Comment by OlegKi** [comment]
If I correctly understand your requirements the you can do small modification in the previous code:
```JavaScript
onSelectAll: function (rowids, toCheck) {
    $(this).find("input.groupHeader").prop("checked", toCheck);
},
loadComplete: function () {
    var $self = $(this);
    $self.find("input.groupHeader").change(function (e) {
        var $tr;
        for ($tr = $(e.target).closest(".jqgroup").next("tr"); $tr.hasClass("jqgrow"); $tr = $tr.next()) {
            $self.jqGrid("setSelection", $tr.attr("id"), true);
        }
    });
}
```
see https://jsfiddle.net/OlegKi/27hmroup/

**Comment by tsbalaji69** [comment]
Hi

**Comment by OlegKi** [comment]
Could you describe the problem more detailed as "Hi"? The main text of your last comment were not included because of some reasons.

**Comment by tsbalaji69** [comment]
Hi OlegKi,
I opened a new issue [https://github.com/free-jqgrid/jqGrid/issues/455](url) as I thought the existing issue to be closed. The above mentioned details helped me to check uncheck all the detail checkbox when a header check box is checked. Thank you. But when any one detail check box is unchecked the header checkbox should be unchecked this is not happening. Attaching a screen shot
![image](https://user-images.githubusercontent.com/43889770/47713108-73f15600-dc5f-11e8-98e3-f8e592b30fbe.png)
Also can we display values for few columns  along with the header row?
Thank you,
regards,
balaji

**Comment by OlegKi** [comment]
Hi,

the previous demo can be easy modified to change the behavior of selection/deselection. One need just test additionally whether the current line already has the selection state, which one needs or not. For example,
```JavaScript
loadComplete: function () {
    var $self = $(this),
        p = $self.jqGrid("getGridParam");

    $self.find("input.groupHeader").change(function (e) {
        var $tr, rowid, toSelect = $(e.target).prop("checked"), isSelected;

        for ($tr = $(e.target).closest(".jqgroup").next("tr"); $tr.hasClass("jqgrow"); $tr = $tr.next()) {
            rowid = $tr.attr("id");
            isSelected = $.inArray(rowid, p.selarrrow) >= 0;
            if ((toSelect && !isSelected) || (!toSelect && isSelected)) {
                $self.jqGrid("setSelection", $tr.attr("id"), true);
            }
        }
    });
}
```
See https://jsfiddle.net/OlegKi/0kL9sf5r/

**Comment by tsbalaji69** [comment]
Hi OlegKi,
The check box present at the header level is not getting unchecked if any one of the detail checkbox is unchecked.
![image](https://user-images.githubusercontent.com/43889770/47735703-11b24880-dc93-11e8-8b71-a4aeeb1d716e.png)
Say for example user checks the header level check box and all the detail level check box is checked. Now the user un checks any detail level check box then the header level check box should also get unchecked as it is not logical to have the header level check box checked and detail level check box un checked.
Hope i described the issue well.

Thank you,
Regards,
Balaji

**Comment by OlegKi** [comment]
I'v shown you in previous demo that rows with data has `jqgrow` class and the row with group header has the class `jqgroup`. You can use `jQuery.next` or `jQuery.prev` to go till the grouping header or to work through all rows in the group.

The next demo https://jsfiddle.net/OlegKi/ay9os4e2/ shows how to deselect the checkbox of the group header if an item from the group will be deselected. I think that you can implement yourself any other your specific requirements in the same way.

**Comment by tsbalaji69** [comment]
Hi OlegKi,
Thank you for the quick response. I will work this out and let you know.
Regards,
Balaji

**Comment by tsbalaji69** [comment]
Hi OlegKi,
Thank you for the guidance. In the same jqgrid when I delete the rows selected I have a requirement to get the selected data and post it to server. I have many groups in multiple pages. user may select them at the header level or at the detail level. I need to distinguish if the user has selected the header level check box and collect all the row values and also at the detail level rows. So I tried to loop the available groups like this
~~~
   var $grid = $("#dataGrid"),i,groups = $grid.jqGrid("getGridParam", "groupingView").groups,
                        l = groups.length, datatosend = [];
     //   alert(l);
        var $groupHeaderRows = $grid.find(">tbody>tr.jqgroup");
        var isHeaderSelected = $groupHeaderRows.prop("checked");
      //  console.log($groupHeaderRows);
        alert(isHeaderSelected);
 //Check if header row is checked push the state to array
     datatosend.push({"HeaderChecked:"+isHeaderSelected+"});
//now loop through the rows under this header and push the selected row data to array
        for ($tr = $groupHeaderRows.next("tr") ; $tr.hasClass("jqgrow") ; $tr = $tr.next()) {
            rowid = $tr.attr("id");
           selectedrowdetails = $grid.jqGrid('getRowData', rowid);
             datatosend.push(selectedrowdetails);
        }

~~~ 
How do we find if the header is checked? .prop("checked"); is not defined. Is this good approach or we can loop through groups and get the checked state and its rows in an efficient way?
example screen shot
![image](https://user-images.githubusercontent.com/43889770/48571451-e3886600-e92c-11e8-8a42-b896c6547020.png)
Thank you,
regards,
balaji

**Comment by OlegKi** [comment]
Hello balaji,

sorry, but I have my main job and my customers. I can't implement all your requirements and requirements of other developers, who use free jqGrid over the world. I wrote you before that you should implement yourself all your specific requirements.

About you problem: I don't understand why you need to get information about groups selected, because the user can deselect some rows after the group is selected in via click on the group header. In the way selection some group gives you practically no additional information. In any way, you can **track the clicks** on the checkboxes of the group headers inside of the corresponding `change` event handler (see `$self.find("input.groupHeader").change(function (e) {...}` in the demo). In the way you will don't need to enumerate the groups for getting the check state of the group.

One advice to you: you can consider to use `multiPageSelection: true` option (see [the answer](https://stackoverflow.com/a/33021115/315935)). As the result the user can select rows on different pages and the selected rows will be not reset on changing the page. You can get the array of selected rowids from `selarrrow` option of jqGrid (`$("#dataGrid").jqGrid("getGridParam", "selarrrow"`). To get the data of any selected row you can use `getLocalRow` method.

**Comment by tsbalaji69** [comment]
Hi OlegKi,
I am very sorry for bothering you. What ever you have guided so far is very much of use and I am thankful for the same. Forgive me if my questions are irritating you and took your time. I sincerely apologize for any inconvenience caused to you. Once again thank you for the needful.
Regards,
Balaji

**Comment by OlegKi** [comment]
I think that your original issue is solved. Please create new issue if you post new problem. I'm closing the issue.
